### PR TITLE
docs: describe the new network configuration subsystem

### PR DESCRIPTION
**Pull request by smira**
Internal details, resources, examples inspecting the configuration.

Signed-off-by: Andrey Smirnov <smirnov.andrey@gmail.com>

### CODE REVIEW

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+NODE         NAMESPACE   TYPE            ID                                       VERSION   ADDRESS                        LINK
+172.20.0.2   network     AddressStatus   eth0/172.20.0.2/24                       1         172.20.0.2/24                  eth0
+172.20.0.2   network     AddressStatus   eth0/fe80::9804:17ff:fe9d:3058/64        2         fe80::9804:17ff:fe9d:3058/64   eth0
+172.20.0.2   network     AddressStatus   flannel.1/10.244.4.0/32                  1         10.244.4.0/32                  flannel.1
+172.20.0.2   network     AddressStatus   flannel.1/fe80::10b5:44ff:fe62:6fb8/64   2         fe80::10b5:44ff:fe62:6fb8/64   flannel.1
+172.20.0.2   network     AddressStatus   lo/127.0.0.1/8                           1         127.0.0.1/8                    lo
+172.20.0.2   network     AddressStatus   lo/::1/128                               1         ::1/128                        lo
+```
+
+In the output there are addresses were set up by Talos (e.g. `eth0/172.20.0.2/24`) and
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:58]:
```suggestion
In the output there are addresses set up by Talos (e.g. `eth0/172.20.0.2/24`) and
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+addresses set up by other facilities (e.g. `flannel.1/10.244.4.0/32` set up by CNI).
+
+Talos networking configuration controllers watch the kernel state and update resources
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:61]:
```suggestion
Talos networking controllers watch the kernel state and update resources
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+```sh
+$ talosctl get resolvers
+NODE         NAMESPACE   TYPE             ID          VERSION   RESOLVERS
+172.20.0.2   network     ResolverStatus   resolvers   2         ["8.8.8.8","1.1.1.1"]
+```
+
+## Inspecting Configuration
+
+Desired networking configuration is combined from multiple sources and presented
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:102]:
```suggestion
The desired networking configuration is combined from multiple sources and presented
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+```sh
+$ talosctl get addressspecs
+NODE         NAMESPACE   TYPE          ID                   VERSION
+172.20.0.2   network     AddressSpec   eth0/172.20.0.2/24   2
+172.20.0.2   network     AddressSpec   lo/127.0.0.1/8       2
+172.20.0.2   network     AddressSpec   lo/::1/128           2
+```
+
+These `AddressSpecs` are applied to the Linux kernel to reach the desired state.
+If, for example, some `AddressSpec` is removed, the address is removed from the Linux network interface as well.
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:114]:
```suggestion
If, for example, an `AddressSpec` is removed, the address is removed from the Linux network interface as well.
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+spec:
+    address: 172.20.0.2/24
+    linkName: eth0
+    family: inet4
+    scope: global
+    flags: permanent
+    layer: operator
+```
+
+One of the important fields is the `layer: operator`, which describes a configuration layer this spec is coming from: in this case, it's generated by a network operator (see below) and is set by the DHCPv4 operator.
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:144]:
```suggestion
An important field is the `layer` field, which describes a configuration layer this spec is coming from: in this case, it's generated by a network operator (see below) and is set by the DHCPv4 operator.
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+
+## Configuration Merging
+
+Spec resources described in the previous section show the final merged configuration state,
+while initial specs are put to a different unmerged namespace `network-config`.
+Spec resources in the `network-config` namespace are merged with conflict resolution to produce the final merged representation in the `network` namespace.
+
+Let's take `HostnameSpec` as an example.
+Final merged representation is:
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:153]:
```suggestion
The final merged representation is:
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+    updated: 2021-06-29T20:23:18Z
+spec:
+    hostname: talos-default-master-1
+    domainname: ""
+    layer: operator
+```
+
+We can see that there are two specs for the hostname:
+
+* one from configuration layer `default` specifies hostname `talos-172-20-0-2` (default driven by the default node address);
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:222]:
```suggestion
* one from the `default` configuration layer which defines the hostname as `talos-172-20-0-2` (default driven by the default node address);
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+* another one from the layer `operator` specifies hostname `talos-default-master-1` (DHCP).
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:223]:
```suggestion
* another one from the layer `operator` that defines the hostname as `talos-default-master-1` (DHCP).
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+
+Talos merges those two specs into final `HostnameSpec` based on the configuration layer and merge rules.
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:225]:
```suggestion
Talos merges these two specs into a final `HostnameSpec` based on the configuration layer and merge rules.
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+Configuration layers in the order of precedence from low to high:
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:226]:
```suggestion
Here is the order of precedence from low to high:
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+
+* `default` (defaults provided by Talos);
+* `cmdline` (from the kernel command line);
+* `platform` (driven by the cloud provider);
+* `operator` (various dynamic config: DHCP, Virtual IP, etc);
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:231]:
```suggestion
* `operator` (various dynamic configuration options: DHCP, Virtual IP, etc);
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+* `configuration` (derived from the machine configuration).
+
+So in our example `operator` layer `HostnameSpec` overwrites `default` layer producing final hostname `talos-default-master-1`.
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:234]:
```suggestion
So in our example the `operator` layer `HostnameSpec` overwrites the `default` layer producing the final hostname `talos-default-master-1`.
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 13]
```diff
@@ -0,0 +1,392 @@
...
+
+The merge process applies to all six core networking specs.
+For each spec `layer` controls the merge behavior, if multiple configuration specs
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:237]:
```suggestion
For each spec, the `layer` controls the merge behavior
If multiple configuration specs
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 14]
```diff
@@ -0,0 +1,392 @@
...
+    operator: dhcp4
+    linkName: eth0
+    requireUp: true
+    dhcp4:
+        routeMetric: 1024
+```
+
+`OperatorSpec` resources are generated by Talos based on machine configuration mostly.
+DHCP4 operator is created automatically for all physical network links which are not configured explicitly via the kernel command line or the machine configuration.
+This also means that Talos on the first boot without a machine configuration attempts DHCP on all physical network interfaces by default.
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:278]:
```suggestion
This also means that on the first boot, without a machine configuration, a DHCP request is made on all physical network interfaces by default.
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 15]
```diff
@@ -0,0 +1,392 @@
...
+
+Specs generated by operators are prefixed with operator ID (`dhcp4/eth0` in the example above) in the unmerged `network-config` namespace:
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:280]:
```suggestion
Specs generated by operators are prefixed with the operator ID (`dhcp4/eth0` in the example above) in the unmerged `network-config` namespace:
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 16]
```diff
@@ -0,0 +1,392 @@
...
+$ talosctl -n 172.20.0.2 get addressspecs --namespace network-config
+NODE         NAMESPACE        TYPE          ID                              VERSION
+172.20.0.2   network-config   AddressSpec   dhcp4/eth0/eth0/172.20.0.2/24   1
+```
+
+## Other Network Resources
+
+There are some additional resources describing the network subsystem state.
+
+`NodeAddress` resource presents node addresses excluding link-local and loopback addresses:
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:292]:
```suggestion
The `NodeAddress` resource presents node addresses excluding link-local and loopback addresses:
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 17]
```diff
@@ -0,0 +1,392 @@
...
+
+```sh
+$ talosctl get nodename
+NODE          NAMESPACE      TYPE       ID         VERSION   NODENAME
+10.100.2.23   controlplane   Nodename   nodename   1         infra-green-cp-mmf7v
+```
+
+Depending on the machine configuration `nodename` might be just a hostname or the FQDN of the node.
+
+`NetworkStatus` aggregates current state of the network configuration:
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:318]:
```suggestion
`NetworkStatus` aggregates the current state of the network configuration:
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 18]
```diff
@@ -0,0 +1,392 @@
...
+spec:
+    addressReady: true
+    connectivityReady: true
+    hostnameReady: true
+    etcFilesReady: true
+```
+
+## Network Controllers
+
+For each of six basic resource types, there are several controllers:
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:341]:
```suggestion
For each of the six basic resource types, there are several controllers:
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 19]
```diff
@@ -0,0 +1,392 @@
...
+
+* `*StatusController` populates `*Status` resources observing Linux kernel state.
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:343]:
```suggestion
* `*StatusController` populates `*Status` resources observing the Linux kernel state.
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 20]
```diff
@@ -0,0 +1,392 @@
...
+* `*ConfigController` produces initial unmerged `*Spec` resources in the `network-config` namespace based on defaults, kernel command line, machine configuration.
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:344]:
```suggestion
* `*ConfigController` produces the initial unmerged `*Spec` resources in the `network-config` namespace based on defaults, kernel command line, and machine configuration.
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 21]
```diff
@@ -0,0 +1,392 @@
...
+
+* `lo` interface is assigned addresses `127.0.0.1/8` and `::1/128`;
+* hostname is set to the `talos-<IP>` where `IP` is the default node address;
+* resolvers are set to `8.8.8.8`, `1.1.1.1`;
+* time servers are set to `pool.ntp.org`;
+* DHCP4 operator is run on any physical interface which is not configured explicitly.
+
+### Cmdline
+
+Kernel command line is parsed for the following options:
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:367]:
```suggestion
The kernel command line is parsed for the following options:
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 22]
```diff
@@ -0,0 +1,392 @@
...
+
+* `ip=` option is parsed for node IP, default gateway, hostname, DNS servers, NTP servers;
+* `talos.hostname=` option is used to set node hostname;
+* `talos.network.interface.ignore=` can be used to make Talos skip network interface configuration completely.
+
+### Platform
+
+Platform configuration delivers cloud environment-specific options, for example, hostname.
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:375]:
```suggestion
Platform configuration delivers cloud environment-specific options (e.g. the hostname).
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 23]
```diff
@@ -0,0 +1,392 @@
...
+
+### Operator
+
+Network operators provide configuration for all basic resource types.
+
+### Machine Configuration
+
+Machine configuration is parsed for link configuration, addresses, routes, hostname,
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:383]:
```suggestion
The machine configuration is parsed for link configuration, addresses, routes, hostname,
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 24]
```diff
@@ -0,0 +1,392 @@
...
+resolvers and time servers.
+Any changes to `.machine.network` configuration can be applied in immediate mode.
+
+## Network Configuration Debugging
+
+Most of the network controller operations and failures are logged to the kernel console,
+additional logs with `debug` level are available with `talosctl logs controller-runtime` command.
+If the networking configuration can't be established and API is not available, `debug` level
```
Comment (andrewrynhard) [inline website/content/docs/v0.11/Learn More/networking-resources.md:391]:
```suggestion
If the networking configuration can't be established and the API is not available, `debug` level
```

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 25]
```diff
@@ -0,0 +1,392 @@
+---
+title: "Networking Resources"
+weight: 10
+---
+
+Talos starting with version 0.11 provides a new implementation of the network configuration subsystem powered by [COSI](../controllers-resources/).
+The new implementation is still using the same machine configuration file format and external sources to configure a node network, so there should be no difference
+in the way Talos works in 0.11.
+
+The most notable change in Talos 0.11 is that all changes to machine configuration `.machine.network` can be applied now in immediate mode (without a reboot) via `talosctl edit mc --immediate`.
```
[the remark this answers is not in the document]
Comment (smira) [reply website/content/docs/v0.11/Learn More/networking-resources.md:10]: yep, we should, I will add it, thanks

File: website/content/docs/v0.11/Learn More/networking-resources.md [hunk 26]
```diff
@@ -0,0 +1,392 @@
...
+| `TimeServer`       | `TimeServerStatus`     | `TimeServerSpec`     |
+
+Status resources have aliases with `Status` part removed, so for example
+`AddressStatus` is also avaiable as `Address`.
+
+Talos networking controllers reconcile the state so that `*Status` equals the desired `*Spec`.
+
+## Observing State
+
+Current networking configuration state can be observed by querying `*Status` resources via
```
[the remark this answers is not in the document]
Comment (smira) [reply website/content/docs/v0.11/Learn More/networking-resources.md:44]: I was torn on whether it should be `networking controllers` vs. `network configuration controllers` or even `network controllers`, as these controllers don't do any networking, they just really set up configuration, but I agree in this sentence it sounds weird

**Review by andrewrynhard — APPROVED** [review]

**Pull request merged**